### PR TITLE
Import certificates

### PR DESCRIPTION
**Pull request by clebs**
**Description**

Changes proposed in this pull request:

- The install command now automatically imports the kyma root certificate
  - Works seamlessly on OSX and Linux
  - Works seamlessly on Windows if on a command line with admin rights
  - Not possible on Windows if the command line has no admin rights => Instructions are printed.
  - Manual certificate import instructions printed in case of failure or user skips certificate import.
- Added mocks for testing
- Added unit tests for the new functionality

**Related issue(s)**
Part of #81

### CODE REVIEW

File: internal/trust/trust_darwin.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+	return decodedCert, nil
+}
+
+func (k keychain) StoreCertificate(file string, i Informer) error {
+	i.LogInfo("Kyma wants to add its root certificate to the keychain.")
+	if root.IsWithSudo() {
+		i.LogInfo("You're running CLI with sudo. CLI has to add the Kyma certificate to the keychain. Type 'y' to allow this action.")
+		if !root.PromptUser() {
+			i.LogInfo(fmt.Sprintf("\nCould not import the kyma root certificate, please follow the instructions below to import it manually:\n-----\n%s-----\n", k.Instructions()))
```
Comment (tomekpapiernik) [inline internal/trust/trust_darwin.go]:
```suggestion
			i.LogInfo(fmt.Sprintf("\nCould not import the Kyma root certificate, please follow the instructions below to import it manually:\n-----\n%s-----\n", k.Instructions()))
```

File: internal/trust/trust_darwin.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+			return nil
+		}
+	}
+
+	_, err := internal.RunCmd("sudo", "security", "add-trusted-cert", "-d", "-r", "trustRoot", "-k", "/Library/Keychains/System.keychain", file)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("\nCould not import the kyma root certificate, please follow the instructions below to import it manually:\n-----\n%s-----\n", k.Instructions()))
```
Comment (tomekpapiernik) [inline internal/trust/trust_darwin.go]:
```suggestion
		return errors.Wrap(err, fmt.Sprintf("\nCould not import the Kyma root certificate, please follow the instructions below to import it manually:\n-----\n%s-----\n", k.Instructions()))
```

File: internal/trust/trust_linux.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+	return decodedCert, nil
+}
+
+func (c certauth) StoreCertificate(file string, i Informer) error {
+	i.LogInfo("Kyma wants to add its root certificate to the trusted certificate store.")
+	if root.IsWithSudo() {
+		i.LogInfo("You're running CLI with sudo. CLI has to add the Kyma certificate to the trusted certificate store. Type 'y' to allow this action.")
+		if !root.PromptUser() {
+			i.LogInfo(fmt.Sprintf("\nCould not import the kyma root certificate, please follow the instructions below to import it manually:\n-----\n%s-----\n", c.Instructions()))
```
Comment (tomekpapiernik) [inline internal/trust/trust_linux.go]:
```suggestion
			i.LogInfo(fmt.Sprintf("\nCould not import the Kyma root certificate, please follow the instructions below to import it manually:\n-----\n%s-----\n", c.Instructions()))
```

File: internal/trust/trust_linux.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+		return err
+	}
+
+	_, err = internal.RunCmd("sudo", "cp", file, fmt.Sprintf("/usr/local/share/ca-certificates/kyma-%s.crt", domain))
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("\nCould not import the kyma certificates, please follow the instructions below to import them manually:\n-----\n%s-----\n", c.Instructions()))
+	}
+	_, err = internal.RunCmd("sudo", "update-ca-certificates")
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("\nCould not import the kyma certificates, please follow the instructions below to import them manually:\n-----\n%s-----\n", c.Instructions()))
```
Comment (tomekpapiernik) [inline internal/trust/trust_linux.go]:
```suggestion
		return errors.Wrap(err, fmt.Sprintf("\nCould not import the Kyma certificates, please follow the instructions below to import them manually:\n-----\n%s-----\n", c.Instructions()))
```

File: internal/trust/trust_linux.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	}
+
+	return nil
+}
+
+func (certauth) Instructions() string {
+	return "1. Download the certificate: kubectl get configmap net-global-overrides -n kyma-installer -o jsonpath='{.data.global\\.ingress\\.tlsCrt}' | base64 --decode > kyma.crt\n" +
+		"2. Rename the certificate file: mv kyma.crt {NAME}\n" +
```
Comment (tomekpapiernik) [inline internal/trust/trust_linux.go]:
```suggestion
		"2. Rename the certificate file: mv kyma.crt {NEW_CERT_NAME}\n" +
```

Comment (tomekpapiernik) [reply internal/trust/trust_linux.go]: I think this helps understand that here you define the new filename. I like being explicit ;)

File: internal/trust/trust_linux.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+		"3. Copy the certificate to the CA folder: sudo cp {NAME} /usr/local/share/ca-certificates/\n" +
```
Comment (tomekpapiernik) [inline internal/trust/trust_linux.go]:
```suggestion
		"3. Copy the certificate to the CA folder: sudo cp {NEW_CERT_NAME} /usr/local/share/ca-certificates/\n" +
```

Comment (tomekpapiernik) [reply internal/trust/trust_linux.go]: Again used `{NEW_CERT_NAME}` as you're working with the renamed certificate file.

File: internal/trust/trust_windows.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	return decodedCert, nil
+}
+
+func (c certutil) StoreCertificate(file string, i Informer) error {
+	i.LogInfo("Kyma wants to add its root certificate to the trusted certificates.")
+
+	if root.IsWithSudo() {
+		i.LogInfo("You're running CLI with sudo. CLI has to add the Kyma root certificate to the trusted certificates. Type 'y' to allow this action.")
+		if !root.PromptUser() {
+			i.LogInfo(fmt.Sprintf("\nCould not import the kyma root certificate, please follow the instructions below to import it manually:\n-----\n%s-----\n", c.Instructions()))
```
Comment (tomekpapiernik) [inline internal/trust/trust_windows.go]:
```suggestion
			i.LogInfo(fmt.Sprintf("\nCould not import the Kyma root certificate, please follow the instructions below to import it manually:\n-----\n%s-----\n", c.Instructions()))
```

File: internal/trust/trust_windows.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+			return nil
+		}
+		// Only automatically add the cert if already on admin mode, can't ask for admin password from go
+		_, err := internal.RunCmd("certutil", "-addstore", "-f", "Root", file)
+		return err
+	}
+	return errors.New(fmt.Sprintf("Could not import the kyma root certificate, please follow the instructions below to import them manually:\n-----\n%s-----\n", c.Instructions()))
```
Comment (tomekpapiernik) [inline internal/trust/trust_windows.go]:
```suggestion
	return errors.New(fmt.Sprintf("Could not import the Kyma root certificate, please follow the instructions below to import them manually:\n-----\n%s-----\n", c.Instructions()))
```

File: internal/trust/trust.go [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+	// StoreCertificate imports the given certificate file into the trusted root certificates manager of the OS.
+	StoreCertificate(file string, info Informer) error
+
+	// Instructions provides instructions on how to manually store a certificate.
+	// Use in case it can not be stored by calling StoreCertificate.
+	Instructions() string
+}
+
+// informer defines the war ceertification management informs about its progress.
```
Comment (tomekpapiernik) [inline internal/trust/trust.go]: I know it's only a comment but something seems off - is this sentence correct?

File: internal/trust/trust_darwin.go [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+func NewCertifier(verbose bool) Certifier {
+	return keychain{
+		kubectl: kubectl.NewWrapper(verbose),
+	}
+}
+
+func (k keychain) Certificate() ([]byte, error) {
+	cert, err := k.kubectl.RunCmd("get", "configmap", "net-global-overrides", "-n", "kyma-installer", "-o", "jsonpath='{.data.global\\.ingress\\.tlsCrt}'")
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("\nCould not obtain the kyma root certificate, please follow the instructions below to import it manually:\n-----\n%s-----\n", k.Instructions()))
```
Comment (tomekpapiernik) [inline internal/trust/trust_darwin.go]:
```suggestion
		return nil, errors.Wrap(err, fmt.Sprintf("\nCould not obtain the Kyma root certificate, please follow the instructions below to import it manually:\n-----\n%s-----\n", k.Instructions()))
```

File: internal/trust/trust_darwin.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	}
+
+	decodedCert, err := base64.StdEncoding.DecodeString(cert)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("\nCould not obtain the kyma root certificate, please follow the instructions below to import it manually:\n-----\n%s-----\n", k.Instructions()))
```
Comment (tomekpapiernik) [inline internal/trust/trust_darwin.go]:
```suggestion
		return nil, errors.Wrap(err, fmt.Sprintf("\nCould not obtain the Kyma root certificate, please follow the instructions below to import it manually:\n-----\n%s-----\n", k.Instructions()))
```

File: internal/trust/trust_linux.go [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+	// get domain to put on the certificate name.
+	// Linux does not have a proper certificate manager and we need to be able to identify the certificate
+	domain, err := certDomain(file)
+	if err != nil {
+		return err
+	}
+
+	_, err = internal.RunCmd("sudo", "cp", file, fmt.Sprintf("/usr/local/share/ca-certificates/kyma-%s.crt", domain))
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("\nCould not import the kyma certificates, please follow the instructions below to import them manually:\n-----\n%s-----\n", c.Instructions()))
```
Comment (tomekpapiernik) [inline internal/trust/trust_linux.go]:
```suggestion
		return errors.Wrap(err, fmt.Sprintf("\nCould not import the Kyma certificates, please follow the instructions below to import them manually:\n-----\n%s-----\n", c.Instructions()))
```

File: internal/trust/trust.go (hunk 9 above)
Comment (clebs) [reply internal/trust/trust.go]: True, there are a couple of typos in the godoc of this interface, thanks for catching that one.

File: internal/trust/trust_darwin.go (hunk 1 above)
Comment (clebs) [reply internal/trust/trust_darwin.go]: Fixed thanks!

File: internal/trust/trust_linux.go (hunk 3 above)
Comment (clebs) [reply internal/trust/trust_linux.go]: Fixed thanks!

File: internal/trust/trust_linux.go (hunk 4 above)
Comment (clebs) [reply internal/trust/trust_linux.go]: Fixed thanks!

File: internal/trust/trust_linux.go (hunk 5 above)
Comment (clebs) [reply internal/trust/trust_linux.go]: Fixed thanks!

File: internal/trust/trust_linux.go (hunk 6 above)
Comment (clebs) [reply internal/trust/trust_linux.go]: Fixed thanks!

File: internal/trust/trust_windows.go (hunk 7 above)
Comment (clebs) [reply internal/trust/trust_windows.go]: Fixed thanks!

File: internal/trust/trust.go (hunk 9 above)
Comment (clebs) [reply internal/trust/trust.go]: Fixed thanks!

File: internal/trust/trust_darwin.go (hunk 11 above)
Comment (clebs) [reply internal/trust/trust_darwin.go]: Fixed thanks!

File: internal/trust/trust_linux.go (hunk 12 above)
Comment (clebs) [reply internal/trust/trust_linux.go]: Fixed thanks!

File: internal/trust/trust_windows.go (hunk 8 above)
Comment (clebs) [reply internal/trust/trust_windows.go]: Fixed thanks!

**Pull request merged**